_episodes/04-drawing-bitwise.md

```diff
@@ -45,103 +46,90 @@ upper-left coordinate of the sub-area we are interested in is *(44, 357)*,
-the alias saves us a little typing. Then, we load and display the initial image 
-in the same way we have done before.
+As before, we first import `skimage`. We also import the NumPy library, and give
+it an alias of `np`. NumPy is necessary when we create the initial mask image,
+and the alias saves us a little typing. Then, we load and display the initial
+image in the same way we have done before.
 
-If you recall our discussion of the RGB color model in the 
-[Image Basics]({{ page.root }}/02-image-basics) lesson, you will remember that
-the color black has RGB values of *(0, 0, 0)*. It follows, then, that an image
-that is all black would be a three-dimensional NumPy array containing nothing
-but zeros. Luckily, the NumPy library provides a function to create just such
-an array. The next section of code shows how.
+NumPy allows indexing of images/arrays with "boolean" arrays of the same size.
+Indexing with a boolean array is also called mask indexing. The "pixels" in such
+a mask array can only take two values: `True` or `False`. When indexing an image
+with such a mask, only pixel values at positions where the mask is `True` are
+accessed. But first, we need to generate a mask array of the same size as the
+image. Luckily, the NumPy library provides a function to create just such an
+array. The next section of code shows how.
 
 ~~~
-# Create the basic black image 
-mask = np.zeros(shape = image.shape, dtype = "uint8")
+# Create the basic mask 
+mask = np.ones(shape=image.shape[0:2], dtype="bool")
 ~~~
 {: .python}
 
-We create the array / all black image with the 
+We create the mask image with the 
 
-`mask = np.zeros(shape = img.shape, dtype = "uint8")`
+`mask = np.ones(shape=image.shape[0:2], dtype="bool")`
 
-function call. The first argument to the `zeros()` function is the shape of 
+function call. The first argument to the `ones()` function is the shape of 
 the original image, so that our mask will be exactly the same size as the 
-original. The second argument, `dtype = "uint8"`, indicates that the elements
-in the array should be 8-bit, unsigned integers -- i.e., numbers with values
-in the range *[0, 255]*. That is exactly what we need for storing RGB values 
-in 24-bit color.
+original. Notice, that we have only used the first two indices of our shape. We
+omitted the channel dimension. Indexing with such a mask will change all channel
+values simultaneously. The second argument, `dtype = "bool"`, indicates that the
+elements in the array should be booleans -- i.e., values are either `True` or
+`False`.
 
-Next, we draw a filled, white rectangle on the all-black mask:
+Next, we draw a filled, rectangle on the mask:
 
 ~~~
-# Draw a white, filled rectangle on the mask image
-cv2.rectangle(img = mask, pt1 = (44, 357), pt2 = (720, 740), 
-	color = (255, 255, 255), thickness = -1)
+# Draw filled rectangle on the mask image
+rr, cc = skimage.draw.rectangle(start=(357, 44), end=(740, 720))
```

Comment:
@k-dominik I think the spatial axis conventions are the same in `cv2` and `skimage`. Hence you don't need to flip axes here.
I haven't checked this myself though, so correct me if you are sure this is right.

Reply:
* openCV has `Point_ (_Tp _x, _Tp _y)`
* sklimage has `r, c int : Centre coordinate of circle.`

I do admit that I did the old trial and error once in the beginning ;)

Reply:
If I understand this correctly, the API of the OpenCV functions and indexing numpy arrays is not consistent?! I.e. OpenCV functions take `(x, y)` whereas the arrays are indexed via `(y, x)`.

Reply:
yep, this is not so overly surprising, as it is natural to give coordinates in this order... Still a confusing choice, another great reason to switch ;)

Reply:
I see. Very confusing from a user perspective indeed...